src/main/scala/rise/Cuda/TypedDSL.scala

```diff
@@ -0,0 +1,51 @@
+package rise.Cuda
+
+import rise.core.DSL._
+import rise.core.semantics.HalfData
+import rise.core.{DepApp, Expr, Literal, Primitive}
+import rise.core.types.Nat
+
+object TypedDSL {
+  object mapBlock {
+    def apply(): ToBeTyped[Primitive] = primitives.mapBlock('x')
+    def apply[T <: Expr](e: ToBeTyped[T]): ToBeTyped[rise.core.App] = primitives.mapBlock('x')(e)
+    def apply(dim: Char): ToBeTyped[Primitive] = primitives.mapBlock(dim)
+  }
+
+  object mapGlobal {
+    def apply(): ToBeTyped[Primitive] = primitives.mapGlobal('x')
+    def apply[T <: Expr](e: ToBeTyped[T]): ToBeTyped[rise.core.App] = primitives.mapGlobal('x')(e)
+    def apply(dim: Char): ToBeTyped[Primitive] = primitives.mapGlobal(dim)
+  }
+
+  object mapThreads {
+    def apply(): ToBeTyped[Primitive] = primitives.mapThreads('x')
+    def apply[T <: Expr](e: ToBeTyped[T]): ToBeTyped[rise.core.App] = primitives.mapThreads('x')(e)
+    def apply(dim: Char): ToBeTyped[Primitive] = primitives.mapThreads(dim)
+  }
+
+  object mapWarp {
+    def apply(): ToBeTyped[Primitive] = primitives.mapWarp('x')
+    def apply[T <: Expr](e: ToBeTyped[T]): ToBeTyped[rise.core.App] = primitives.mapWarp('x')(e)
+    def apply(dim: Char): ToBeTyped[Primitive] = primitives.mapWarp(dim)
+  }
+
+  object mapLane {
+    def apply(): ToBeTyped[Primitive] = primitives.mapLane('x')
+    def apply[T <: Expr](e: ToBeTyped[T]): ToBeTyped[rise.core.App] = primitives.mapLane('x')(e)
+    def apply(dim: Char): ToBeTyped[Primitive] = primitives.mapLane(dim)
+  }
+
+  object tensorMMA {
+    def apply[T <: Expr](a: ToBeTyped[T]): ToBeTyped[rise.core.App] = primitives.tensorMMA(a)
+    def apply[T <: Expr, U <: Expr](a: ToBeTyped[T], b: ToBeTyped[U]): ToBeTyped[rise.core.App] = primitives.tensorMMA(a)(b)
+    def apply[T <: Expr, U <: Expr, V <: Expr](a: ToBeTyped[T], b: ToBeTyped[U], c: ToBeTyped[V]): ToBeTyped[rise.core.App] = primitives.tensorMMA(a)(b)(c)
+  }
+
+  object toSharedMemoryShift {
+    def apply(shift: Nat): Expr = primitives.toSharedMemoryShift(shift)
+    def apply(shift: Nat, array: Expr): Expr = primitives.toSharedMemoryShift(shift)(array)
+  }
+
+  def h(f: Float): ToBeTyped[Literal] = literal(HalfData(f))
```

Comment:
can this be moved to the core DSL for now? half floats can be useful in C/OpenCL as well.

Reply:
Yes, i moved it to core DSL. I am not sure if "h" is an good name for this function. The other literals are named "l", but the "l"-function with argument Float is already defined for float-literal...

Reply:
I think that we should have `lf16`, `lf32` and `lf64`, like we have `lu8`, `lu16`, ..

Reply:
https://github.com/rise-lang/shine/blob/b3086143da4a042ade28dbe6d662db43464419ac/src/main/scala/rise/core/DSL/package.scala#L453-L455
You can also consider removing `HalfData` altogether and defining `lf16(v) = cast(lf32(v)) :: f16`